File: app/api/control/gateways.py
"""``/api/v1/gateways`` and ``/api/v1/keys`` — published endpoints and their credentials.

Same three-line shape as the directory and model routes: name a capability, take
:data:`CurrentActor`, hand both to the service.

The capability split is the one thing here that is not uniform, and it is deliberate.
Gateway configuration is ``resources:write``, which an ``org_member`` has — SPEC §5.2 puts
connectors and gateways in their remit. Keys are ``keys:manage``, which they do not, because
a key is a bearer credential for the organization's spend and the task 04 matrix restricts
minting and revoking one to ``org_admin`` and above. Listing keys is only a read, so it
rides on ``org:read``: the response has no secret in it.

``DELETE /keys/{id}`` is a top-level route rather than nested under its gateway. The key id
is unique and already identifies its gateway, and the service scopes it through that
gateway — so a nested path would add a segment the caller has to get right and the server
would have to check for consistency, which is two ways to be wrong instead of none.
"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, Query, status

from app.api.control.deps import CurrentActor, get_gateway_service, require_capability
from app.schemas.common import Page
from app.schemas.gateway import (
    ApiKeyCreateRequest,
    ApiKeyResponse,
    GatewayCreateRequest,
    GatewayResponse,
    GatewayTestRequest,
    GatewayTestResponse,
    GatewayUpdateRequest,
    IssuedApiKeyResponse,
)
from app.services.gateways import GatewayService
from app.services.permissions import Capability

router = APIRouter(tags=["gateways"])

_Service = Annotated[GatewayService, Depends(get_gateway_service)]
_Cursor = Annotated[str | None, Query(max_length=64)]
_Limit = Annotated[int | None, Query(ge=1, le=200)]

_reads = Depends(require_capability(Capability.ORG_READ))
_writes = Depends(require_capability(Capability.RESOURCES_WRITE))
_keys = Depends(require_capability(Capability.KEYS_MANAGE))


@router.get("/gateways", dependencies=[_reads])
async def list_gateways(
    actor: CurrentActor,
    service: _Service,
    cursor: _Cursor = None,
    limit: _Limit = None,
) -> Page[GatewayResponse]:
    page = await service.list_gateways(actor, cursor=cursor, limit=limit)
    return Page(
        items=[GatewayResponse.of(view) for view in page.items],
        next_cursor=page.next_cursor,
    )


@router.post("/gateways", status_code=status.HTTP_201_CREATED, dependencies=[_writes])
async def create_gateway(
    body: GatewayCreateRequest,
    actor: CurrentActor,
    service: _Service,
) -> GatewayResponse:
    return GatewayResponse.of(await service.create_gateway(actor, body.to_draft()))


@router.get("/gateways/{gateway_id}", dependencies=[_reads])
async def get_gateway(
    gateway_id: uuid.UUID,
    actor: CurrentActor,
    service: _Service,
) -> GatewayResponse:
    return GatewayResponse.of(await service.get_gateway(actor, gateway_id))


@router.patch("/gateways/{gateway_id}", dependencies=[_writes])
async def update_gateway(
    gateway_id: uuid.UUID,
    body: GatewayUpdateRequest,
    actor: CurrentActor,
    service: _Service,
) -> GatewayResponse:
    """Partial. The three config sections are deep-merged, so a form that owns one
    section can save without knowing what the others contain.

    ``slug`` is refused rather than ignored — it is part of a URL clients have deployed.
    """
    return GatewayResponse.of(await service.update_gateway(actor, gateway_id, body.to_patch()))


@router.delete(
    "/gateways/{gateway_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[_writes],
)
async def delete_gateway(
    gateway_id: uuid.UUID,
    actor: CurrentActor,
    service: _Service,
) -> None:
    """Takes its keys and targets with it. Every deployed client stops working, which is
    why the UI asks for the slug to be typed."""
    await service.delete_gateway(actor, gateway_id)


@router.post("/gateways/{gateway_id}/test", dependencies=[_writes])
async def test_gateway(
    gateway_id: uuid.UUID,
    body: GatewayTestRequest,
    actor: CurrentActor,
    service: _Service,
) -> GatewayTestResponse:
    """Send a probe completion through the real proxy path.

    Returns the assembled prompt as well as the answer, which is the part that makes this
    a debugging tool rather than a health check: it is the only way to see what the system
    context actually became before task 07's request log exists.
    """
    return GatewayTestResponse.of(
        await service.test_gateway(actor, gateway_id, message=body.message)
    )


# ---------------------------------------------------------------------------
# keys
# ---------------------------------------------------------------------------


@router.get("/gateways/{gateway_id}/keys", dependencies=[_reads])
async def list_keys(
    gateway_id: uuid.UUID,
    actor: CurrentActor,
    service: _Service,
) -> list[ApiKeyResponse]:
    """Not paginated: a gateway holds tens of keys, not thousands, and the screen shows
    them all at once. The cap in the service is what keeps that true."""
    return [ApiKeyResponse.of(key) for key in await service.list_keys(actor, gateway_id)]


@router.post(
    "/gateways/{gateway_id}/keys",
    status_code=status.HTTP_201_CREATED,
    dependencies=[_keys],
)
async def create_key(
    gateway_id: uuid.UUID,
    body: ApiKeyCreateRequest,
    actor: CurrentActor,
    service: _Service,
) -> IssuedApiKeyResponse:
    """The only response in this API containing a usable secret, and the only time this
    one exists. Only ``sha256(secret)`` is stored, so it cannot be shown again."""
    return IssuedApiKeyResponse.of(
        await service.create_key(actor, gateway_id, name=body.name, expires_at=body.expires_at)
    )


@router.delete("/keys/{key_id}", dependencies=[_keys])
async def revoke_key(
    key_id: uuid.UUID,
    actor: CurrentActor,
    service: _Service,
) -> ApiKeyResponse:
    """Soft. The row survives so task 07's request logs keep a reference that resolves,
    and the next request with this key is refused — key lookup is never cached."""
    return ApiKeyResponse.of(await service.revoke_key(actor, key_id))


__all__ = ["router"]

File: web/src/api/gateways.ts
/**
 * Queries and mutations for gateways and their API keys.
 *
 * One module, like `models.ts`, so the cache keys are written next to each other. Two of
 * them are load-bearing:
 *
 * Creating or revoking a **key** invalidates the *gateway* queries as well as the key
 * list, because the list screen shows a key count. Forgetting that is how a page shows
 * "0 keys" next to a key somebody just made.
 *
 * The key mint is a mutation with **no cached result**. Its response is the only place
 * the plaintext exists, so it is handed to the caller and deliberately never written into
 * the query cache — a cache entry is a copy, and a copy of a secret that cannot be shown
 * twice is a copy too many.
 */

import { useMutation, useQuery, useQueryClient, type UseQueryResult } from '@tanstack/react-query'

import { useApiClient } from '@/auth/AuthContext'
import type {
  ApiKeyCreateRequest,
  ApiKeyResponse,
  GatewayCreateRequest,
  GatewayPage,
  GatewayResponse,
  GatewayTestRequest,
  GatewayTestResponse,
  GatewayUpdateRequest,
  IssuedApiKeyResponse,
  MemoryPreviewRequest,
  PromptPreviewResponse,
  RetrievalPreviewResponse,
} from '@/api/types'

export const keys = {
  all: ['gateways'] as const,
  list: (cursor?: string | null) => ['gateways', { cursor: cursor ?? null }] as const,
  one: (id: string) => ['gateways', id] as const,
  apiKeys: (gatewayId: string) => ['gateways', gatewayId, 'keys'] as const,
}

function listPath(cursor?: string | null): string {
  return cursor ? `/api/v1/gateways?cursor=${encodeURIComponent(cursor)}` : '/api/v1/gateways'
}

// -- reads -----------------------------------------------------------------

export function useGateways(cursor?: string | null): UseQueryResult<GatewayPage> {
  const client = useApiClient()
  return useQuery({
    queryKey: keys.list(cursor),
    queryFn: () => client.get<GatewayPage>(listPath(cursor)),
  })
}

export function useGateway(id: string | undefined): UseQueryResult<GatewayResponse> {
  const client = useApiClient()
  return useQuery({
    queryKey: keys.one(id ?? ''),
    queryFn: () => client.get<GatewayResponse>(`/api/v1/gateways/${id}`),
    enabled: Boolean(id),
  })
}

export function useApiKeys(gatewayId: string | undefined): UseQueryResult<ApiKeyResponse[]> {
  const client = useApiClient()
  return useQuery({
    queryKey: keys.apiKeys(gatewayId ?? ''),
    queryFn: () => client.get<ApiKeyResponse[]>(`/api/v1/gateways/${gatewayId}/keys`),
    enabled: Boolean(gatewayId),
  })
}

// -- writes ----------------------------------------------------------------

/**
 * Invalidate every gateway query, not the clever subset.
 *
 * A rename changes the list and the row's own entry; a key change changes the list's key
 * count. Working out which is how a stale count survives a save, and these lists are
 * small.
 */
function useInvalidateGateways(): () => Promise<void> {
  const queryClient = useQueryClient()
  return async () => {
    await queryClient.invalidateQueries({ queryKey: keys.all })
  }
}

export function useCreateGateway() {
  const client = useApiClient()
  const invalidate = useInvalidateGateways()
  return useMutation({
    mutationFn: (body: GatewayCreateRequest) =>
      client.post<GatewayResponse>('/api/v1/gateways', body),
    onSuccess: invalidate,
  })
}

export function useUpdateGateway(id: string | undefined) {
  const client = useApiClient()
  const invalidate = useInvalidateGateways()
  return useMutation({
    mutationFn: (body: GatewayUpdateRequest) =>
      client.patch<GatewayResponse>(`/api/v1/gateways/${id}`, body),
    onSuccess: invalidate,
  })
}

export function useDeleteGateway() {
  const client = useApiClient()
  const invalidate = useInvalidateGateways()
  return useMutation({
    mutationFn: (id: string) => client.delete<void>(`/api/v1/gateways/${id}`),
    onSuccess: invalidate,
  })
}

/**
 * Mint a key. The response carries the only copy of the plaintext there will ever be, so
 * the caller shows it once and lets it go; nothing here retains it.
 */
export function useCreateApiKey(gatewayId: string | undefined) {
  const client = useApiClient()
  const invalidate = useInvalidateGateways()
  return useMutation({
    mutationFn: (body: ApiKeyCreateRequest) =>
      client.post<IssuedApiKeyResponse>(`/api/v1/gateways/${gatewayId}/keys`, body),
    onSuccess: invalidate,
  })
}

export function useRevokeApiKey() {
  const client = useApiClient()
  const invalidate = useInvalidateGateways()
  return useMutation({
    mutationFn: (keyId: string) => client.delete<ApiKeyResponse>(`/api/v1/keys/${keyId}`),
    onSuccess: invalidate,
  })
}

/** Send a probe completion through the real proxy path. */
export function useTestGateway(gatewayId: string | undefined) {
  const client = useApiClient()
  return useMutation({
    mutationFn: (body: GatewayTestRequest) =>
      client.post<GatewayTestResponse>(`/api/v1/gateways/${gatewayId}/test`, body),
  })
}

/**
 * Try retrieval: the chunks a question would inject, with their scores.
 *
 * A mutation rather than a query, and the result is deliberately not cached. It is an
 * action somebody takes — type, press, read — not state the screen is showing, and a
 * cached answer from before the last edit to the score floor is worse than no answer.
 */
export function useTryRetrieval(gatewayId: string | undefined) {
  const client = useApiClient()
  return useMutation({
    mutationFn: (body: MemoryPreviewRequest) =>
      client.post<RetrievalPreviewResponse>(
        `/api/v1/gateways/${gatewayId}/try-retrieval`,
        body,
      ),
  })
}

/** The fully assembled prompt for a sample question, layer by layer. */
export function usePromptPreview(gatewayId: string | undefined) {
  const client = useApiClient()
  return useMutation({
    mutationFn: (body: MemoryPreviewRequest) =>
      client.post<PromptPreviewResponse>(
        `/api/v1/gateways/${gatewayId}/prompt-preview`,
        body,
      ),
  })
}
